spdmlib/src/common/opaque.rs

```diff
@@ -30,81 +26,156 @@ pub const SUPPORTED_VERSION_LIST_SM_DATA_ID: u8 = 0x01;
-];
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub struct GeneralOpaqueDataHeader;
+
+impl Codec for GeneralOpaqueDataHeader {
+    fn encode(&self, bytes: &mut Writer) -> Result<usize, codec::EncodeErr> {
+        let mut cnt = 0;
+
+        cnt += DMTF_SPEC_ID.encode(bytes)?;
+        cnt += DMTF_OPAQUE_VERSION.encode(bytes)?;
+        cnt += OPAQUE_LIST_TOTAL_ELEMENTS.encode(bytes)?;
+        cnt += 0u16.encode(bytes)?; // reserved
+
+        Ok(cnt)
+    }
+
+    fn read(r: &mut Reader) -> Option<Self> {
+        let spec_id = u32::read(r)?;
+        let opaque_version = u8::read(r)?;
+        let opaque_list_total_elements = u8::read(r)?;
+        u16::read(r)?; // reserved
+
+        if spec_id != DMTF_SPEC_ID
+            || opaque_version != DMTF_OPAQUE_VERSION
+            || opaque_list_total_elements != OPAQUE_LIST_TOTAL_ELEMENTS
+        {
+            None
+        } else {
+            Some(Self)
+        }
+    }
+}
+
+impl SpdmCodec for GeneralOpaqueDataHeader {
+    fn spdm_encode(&self, _context: &mut SpdmContext, bytes: &mut Writer) -> SpdmResult<usize> {
+        self.encode(bytes).map_err(|_| SPDM_STATUS_BUFFER_FULL)
+    }
+
+    fn spdm_read(_context: &mut SpdmContext, r: &mut Reader) -> Option<Self> {
+        GeneralOpaqueDataHeader::read(r)
+    }
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub struct FM1OpaqueDataHeader;
+
+impl Codec for FM1OpaqueDataHeader {
+    fn encode(&self, bytes: &mut Writer) -> Result<usize, codec::EncodeErr> {
+        let mut cnt = 0;
+
+        cnt += OPAQUE_LIST_TOTAL_ELEMENTS.encode(bytes)?;
+        cnt += u24::new(0).encode(bytes)?; // reserved
+
+        Ok(cnt)
+    }
+
+    fn read(r: &mut Reader) -> Option<Self> {
+        let opaque_list_total_elements = u8::read(r)?;
+        u24::read(r)?; // reserved
+
+        if opaque_list_total_elements != OPAQUE_LIST_TOTAL_ELEMENTS {
+            None
+        } else {
+            Some(Self)
+        }
+    }
+}
+
+impl SpdmCodec for FM1OpaqueDataHeader {
+    fn spdm_encode(&self, _context: &mut SpdmContext, bytes: &mut Writer) -> SpdmResult<usize> {
+        self.encode(bytes).map_err(|_| SPDM_STATUS_BUFFER_FULL)
+    }
+
+    fn spdm_read(_context: &mut SpdmContext, r: &mut Reader) -> Option<Self> {
+        FM1OpaqueDataHeader::read(r)
+    }
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum SMOpaqueDataHeader {
+    General(GeneralOpaqueDataHeader),
+    FM1(FM1OpaqueDataHeader),
+}
+
+impl Codec for SMOpaqueDataHeader {
+    fn encode(&self, bytes: &mut Writer) -> Result<usize, codec::EncodeErr> {
+        match self {
+            SMOpaqueDataHeader::General(g) => g.encode(bytes),
+            SMOpaqueDataHeader::FM1(f) => f.encode(bytes),
+        }
+    }
+
+    fn read(r: &mut Reader) -> Option<Self> {
+        let buf = r.as_ref();
+        if buf.len() < 4 {
+            None
+        } else {
+            let id = u32::read_bytes(buf)?;
+            if id == DMTF_SPEC_ID {
```

Comment:
I think we need to use SPDM version to determine.

SPDM 1.1 use GeneralOpaqueDataHeader.
SPDM 1.2+ use FM1OpaqueDataHeader

Reply:
> I think we need to use SPDM version to determine.
> 
> SPDM 1.1 use GeneralOpaqueDataHeader. SPDM 1.2+ use FM1OpaqueDataHeader

Suggestion applied